### PR TITLE
Experimental @defer support

### PR DESCRIPTION
**Pull request by TOTBWF**
This PR adds support for the defer directive. Documentation is pretty sparse on it, and there are a couple of behaviours I'd like to work out with everyone (Nested defers, what to do with errors, etc). 

Implementing this necessitated a rewrite of our resolver execution code, because in our old model, our field selection would only give us the leaves of the result tree, and a deferred value could rely on a node.

### CODE REVIEW

**Comment by TOTBWF** [comment]
Keep in mind that we are still waiting on https://github.com/travis-ci/travis-build/pull/1060 and https://github.com/fsprojects/FSharp.Control.Reactive/pull/93 to go through

File: src/FSharp.Data.GraphQL.Server/Execution.fs [hunk 1]
```diff
@@ -368,13 +439,74 @@ let internal executePlan (ctx: ExecutionContext) (plan: ExecutionPlan) (objdef:
...
+                            let! tree' = treeToDict tree
+                            return NameValueLookup.ofList["data", upcast tree'; "path", upcast path]
+                        }) >> AsyncVal.collectParallel))
+                |> AsyncVal.collectParallel
+                |> AsyncVal.map(Array.fold (Array.append) [||])))
+            |> AsyncVal.collectParallel
+            |> AsyncVal.map(Array.fold (Array.append) [||])
+            |> AsyncVal.toAsync
+            |> Observable.ofAsync
+            |> Observable.bind(Observable.ofSeq)
```
Comment (Horusiath) [inline src/FSharp.Data.GraphQL.Server/Execution.fs]: If I read it right, you're basically load a whole collection to be deferred into memory, and `@defer` is just used to return it in chunks as `IObservable<>`? If it's true, I don't see any value added by defer directive.

File: paket.dependencies [hunk 2]
```diff
@@ -1,9 +1,10 @@
 source https://nuget.org/api/v2
-nuget BenchmarkDotNet
-nuget BenchmarkDotNet.Diagnostics.Windows
+nuget BenchmarkDotNet 
+nuget BenchmarkDotNet.Diagnostics.Windows 
 nuget FParsec
 nuget FSCheck
 nuget FSharp.Data.TypeProviders
+nuget FSharp.Control.Reactive
```
Comment (Horusiath) [inline paket.dependencies]: Why do we need it for? `IObservable<>` is part of mscorlib.

**Review by Horusiath — COMMENTED** [review]
We would need some benchmarks for that one. I'm also concerned about the implementation itself, as it looks like trying to prefetch all data upfront with the major data resolution flow, then serving it as Observable, while IMHO it should:

1. Continuously stream data as it's taken from the resolver (so resolver func returns an observable in this case, and maybe new custom type def). 
2. Fetch data aside using async, regardless of how type def has defined resolver - so in this case, deferred fields will always be resolved in parallel to the main resolution flow.

Another thing is that, we should try to bring a more general solution (I've started a discussion on that in #102). Define the requirements and API necessary to aim even such possibilities as deferred fields. But this is something that can eventually be done later and doesn't block this PR from merge.

File: paket.dependencies (hunk 2 above)
Comment (TOTBWF) [reply paket.dependencies]: We need it for some of the transformations from other data types to observables, as well as the `bind` operation on observable

File: src/FSharp.Data.GraphQL.Server/ObservableExtensions.fs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+namespace FSharp.Data.GraphQL
+
+open System
+open System.Reactive
+open System.Reactive.Linq
+
+/// Extension methods to observable, used in place of Fsharp.Control.Reactive
+module Observable =
```
Comment (johnberzy-bazinga) [inline src/FSharp.Data.GraphQL.Server/ObservableExtensions.fs]: We should probably mark this as internal.

**Comment by Horusiath** [comment]
@TOTBWF What is the status on this one?

**Comment by johnberzy-bazinga** [comment]
@Horusiath I think @TOTBWF is just waiting on travis-ci/travis-build#1060. If there is no movement on that PR in the next couple of days we can probably workaround it in the build script, or alternatively, just merge as-is.

**Comment by TOTBWF** [comment]
This should be more or less good to go, there are a few little spots that display some odd behavior sometimes (nested fragment spreads in lists and nested lists of unions), but beyond that it should be more or less complete.

**Comment by TOTBWF** [comment]
Closing this, as it has been subsumed by the subscription support

**Pull request closed without merging**